### PR TITLE
✨(documents): make document context hybrid

### PR DESCRIPTION
**Pull request by camilleAND**
Keep oversized documents tool-call-only, add FIFO context inlining, and align tools/tests with new document payload

## Summary

This PR introduces a hybrid document-context strategy for chat requests.  
Each attached document is represented in the instructions payload and can be either:

- `full-context` (inlined in prompt context), or
- `tool_call_only` (accessible through tools, not inlined).

Inlining is controlled by a token budget with FIFO eviction, so newer documents can be inlined while older ones are downgraded to tool access when needed.

## What Changed

- Added a dedicated document context builder in `chat/document_context_builder.py`:
  - Builds a structured payload with `documents_order` + per-document metadata.
  - Uses `document_id` (UUID) instead of positional `index`.
  - Applies title normalization only for converted files (via `conversion_from`).
  - Supports `full-context` vs `tool_call_only` access modes.
  - Applies FIFO eviction for inlined docs when budget is exceeded.
- Hardened failure behavior in `_load_document`:
  - If `read_attachment_content` fails, document stays `tool_call_only`.
  - Sets a sentinel `token_count` (`sys.maxsize`) to prevent accidental later promotion to `full-context`.
- Integrated context-window controls through configuration/model metadata:
  - `DOCUMENT_CONTEXT_BUDGET_RATIO`
  - `DOCUMENT_CONTEXT_SECURITY_BUFFER_TOKENS`
  - model `max_token_context` support.
- Updated tool behavior:
  - `document_summarize` now supports targeting a single attachment via `document_id`.
  - RAG search path supports document-level filtering (optional metadata filter / fallback behavior).
- Updated prompt/tool descriptions and changelog to reflect the new document-context behavior.
- Added/updated dependency/config surface needed by backend runtime (including `django-solo` in backend deps/settings).

## Why

- Keeps prompt size predictable while preserving access to all attached documents.
- Prevents oversized or unreadable attachments from bloating/poisoning prompt context.
- Improves determinism by using a structured payload (`document_id`, explicit access mode).
- Aligns summarize/search flows with explicit document scoping.

## Test Coverage

- Added/updated context-window tests for:
  - oversized docs remaining `tool_call_only`,
  - FIFO inlining behavior,
  - configurable budget ratio.
- Updated summarize tests for `document_id` selection and queryset-like attachment handling.
- Updated conversation upload/url tests to validate the new instructions payload shape:
  - `document_id` field,
  - converted filename behavior (`*.md`),
  - `info` behavior (`None` vs labeled cases),
  - robust JSON-field assertions instead of brittle full-string reconstruction.

### About Tiktokken
I benchmarked the tiktoken tokenizer against Mistral’s tokenizer on a sample of 10 documents (PDF and Markdown). On average, tiktoken reports about 7% more tokens than Mistral. This overestimation is acceptable for now, and even preferable in our context, since we want to err on the safe side when budgeting document context.

## Summary by CodeRabbit

* **New Features**
  * Target individual attachments via document_id for search and summarize; async search supported. Per-turn document-context listing with token-aware inlining and optional per-model max token context.

* **Behavior**
  * Deterministic newest→oldest ordering, FIFO eviction when budget exceeded, and tool-only marking for oversized/unreadable docs. Document-scoped searches propagate filters, log filtered no-results, and surface retry guidance for empty or invalid IDs.

* **Documentation**
  * Updated docs and tool examples for hybrid context, inlining budget, caching, and document-scoped tools.

* **Tests**
  * Expanded unit/integration tests covering selection, budgeting, caching, retry, and I/O edge cases.

* **Chores**
  * Added runtime token-counting dependency.

### CODE REVIEW

File: src/backend/chat/document_context_builder.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+            "DOCUMENT_CONTEXT_BUDGET_RATIO is 0 for model '%s'; disabling full document inlining.",
+            model_hrid,
+        )
+        payload = _build_tool_only_payload(text_attachments)
+        return (
+            "List of documents attached to this conversation:\n"
+            f"{json.dumps(payload, ensure_ascii=False, indent=2)}"
+        )
+
+    security_buffer = 1000  # Tokens reserved for system instructions and other overhead
```
Comment (providenz) [inline src/backend/chat/document_context_builder.py]: I'd rather have a configurable value

File: src/backend/chat/tools/document_search_rag.py [hunk 2]
```diff
@@ -16,30 +17,59 @@ def add_document_rag_search_tool(agent: Agent) -> None:
...
-        Returns:
-            ToolReturn: Retrieved passages in ``return_value`` and source URLs in
-            ``metadata["sources"]``.
+        When ``document_index`` is provided, search is filtered to a single
+        text attachment (1-based index from the context documents list).
 
         Args:
             ctx (RunContext): The run context containing the conversation.
             query (str): The query to search the documents for.
+            document_index (int | None): Optional document filter, 1-based index
```
Comment (providenz) [inline src/backend/chat/tools/document_search_rag.py]: Not keen of a positional index, it might be fragile, and a 1-based index is error prone.
I'd rather use doc uuid . I'd rather expose the attachment UUID (or a stable slug) in the  jsoon context and have the tool look it up by ID

File: src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py [hunk 3]
```diff
@@ -187,13 +200,24 @@ def search(self, query: str, results_count: int = 4, **kwargs) -> RAGWebResults:
...
                 "k": results_count,  # Number of chunks to return from the search
+                "metadata_filters": metadata_filter,
             },
             timeout=settings.ALBERT_API_TIMEOUT,
         )
         response.raise_for_status()
 
         searches = Searches(**response.json())
 
+        if not searches.data and metadata_filter:
```
Comment (providenz) [inline src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py]: If searches is empty for some reason, the filter is dropped and the tool searches all doc ? is this waht we want ? At least we should log.warning, better if we let the issue surface to the user or maybe use a retry.

File: src/backend/chat/tools/document_summarize.py [hunk 4]
```diff
@@ -79,18 +86,33 @@ async def document_summarize(  # pylint: disable=too-many-locals
...
 
         # Collect documents content
-        text_attachment = await sync_to_async(list)(
-            ctx.deps.conversation.attachments.filter(
-                content_type__startswith="text/",
-            )
+        text_attachment_qs = ctx.deps.conversation.attachments.filter(
+            content_type__startswith="text/"
         )
+        if hasattr(text_attachment_qs, "order_by"):
```
Comment (providenz) [inline src/backend/chat/tools/document_summarize.py]: Why this check ? is this for tests ? if yes we should rather update the test with a mock  supporting order_by

File: src/backend/chat/clients/pydantic_ai.py [hunk 5]
```diff
@@ -683,7 +708,38 @@ def _prepare_prompt(  # noqa: PLR0912  # pylint: disable=too-many-branches
...
+            logger.warning(
+                "Invalid DOCUMENT_CONTEXT_BUDGET_RATIO=%s, falling back to 0.5",
+                budget_ratio,
+            )
+            return 0.5
+        return budget_ratio
+
+    async def _build_document_context_instruction(self) -> str:
+        budget_ratio = self._get_document_context_budget_ratio()
+        text_attachments = await sync_to_async(list)(
```
Comment (providenz) [inline src/backend/chat/clients/pydantic_ai.py:738]: In my understanding this read all files sequentially .It could substancially increase the TFTT. If possible, it would be nice to parallelize the reads (maybe with asyncio.gather).

File: src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py [hunk 6]
```diff
@@ -187,13 +200,24 @@ def search(self, query: str, results_count: int = 4, **kwargs) -> RAGWebResults:
                 "prompt": query,
                 "score_threshold": 0.6,
                 "k": results_count,  # Number of chunks to return from the search
+                "metadata_filters": metadata_filter,
```
Comment (providenz) [inline src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py]: Api accepts none/null adn treat is as "no filters" ?

File: src/backend/chat/clients/pydantic_ai.py [hunk 7]
```diff
@@ -163,6 +165,29 @@
 DOCUMENT_URL_PREFIX = "/media-key/"
 
 
+def count_approx_tokens(text: str) -> int:
+    """Estimate token count using tiktoken."""
+    if not text:
+        return 0
+    try:
+        tiktoken_len = len(tiktoken.get_encoding("cl100k_base").encode(text))
```
Comment (providenz) [inline src/backend/chat/clients/pydantic_ai.py]: we could cache it at the module level : TOKEN_ENCODING = tiktoken.get_encoding("cl100_base")

File: src/backend/chat/document_context_builder.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+
+def _build_tool_only_payload(
+    text_attachments: Sequence[models.ChatConversationAttachment],
+) -> dict:
+    """Build a payload where all documents are accessible via tools only."""
+    docs = [
+        {
+            "index": index,
+            "title": attachment.file_name.removesuffix(".md"),
```
Comment (providenz) [inline src/backend/chat/document_context_builder.py]: What about real md files ? we could check the conversion_from ?

File: src/backend/chat/document_context_builder.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        model_hrid,
+        budget_ratio,
+        document_budget,
+        inlined_total,
+        filled_ratio,
+        len(docs),
+        len(inlined_docs),
+    )
+
+    payload = {
```
Comment (providenz) [inline src/backend/chat/document_context_builder.py]: we have here to divergent payloads to maintain. i'd rather have a single one, possibly typed, with conditional fields

File: src/backend/chat/document_context_builder.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    docs = [
+        {
+            "index": index,
+            "title": attachment.file_name.removesuffix(".md"),
+            "access": "tool_call_only",
+            "content": None,
+            "info": (
+                "first_uploaded_document"
+                if index == 1
+                else ("last_uploaded_document" if index == len(text_attachments) else None)
```
Comment (providenz) [inline src/backend/chat/document_context_builder.py]: what about one attahcment ? probably harmless but looks inconsistent.

File: src/backend/chat/clients/pydantic_ai.py [hunk 11]
```diff
@@ -163,6 +165,29 @@
...
+        logger.debug("Tiktoken length: %s", tiktoken_len)
+        return tiktoken_len
+    except Exception:  # pylint: disable=broad-except #noqa: BLE001
```
Comment (providenz) [inline src/backend/chat/clients/pydantic_ai.py]: We shoud narrow it if possible

File: src/backend/chat/clients/pydantic_ai.py [hunk 12]
```diff
@@ -694,11 +750,12 @@ def summarization_system_prompt() -> str:
...
         @self.conversation_agent.instructions
         def attached_documents_note() -> str:
-            return (
+            base_note = (
                 "[Internal context] User documents are attached to this conversation. "
                 "Do not request re-upload of documents; consider them already available "
-                "via the internal store."
+                "either here in context or via the internal store."
             )
+            return f"{base_note}\n\n{document_context_instruction}"
```
Comment (providenz) [inline src/backend/chat/clients/pydantic_ai.py:795]: When no document_context_instruction, we have trailing line returns

File: src/backend/chat/tests/tools/test_document_search_rag.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+from chat.tools.document_search_rag import add_document_rag_search_tool
+
+
+def test_document_search_rag_schema_accepts_document_index():
+    """The tool schema must expose document_index as an optional argument."""
+    agent = Agent("test")
+    add_document_rag_search_tool(agent)
+
+    schema = agent._function_toolset.tools["document_search_rag"].function_schema.json_schema  # pylint: disable=protected-access
```
Comment (providenz) [inline src/backend/chat/tests/tools/test_document_search_rag.py]: Any way to avoid protected access on pydantic ai ? pretty sure they will break it in an upcoming release

File: src/backend/chat/tools/document_search_rag.py (hunk 2 above)
Comment (camilleAND) [reply src/backend/chat/tools/document_search_rag.py]: Done. I have replaced the use of a 1-based index for the UUID attached to the documents. UUIDs are checked whenever the model use them to prevent hallucinations. I tested for both tools (rag and summarize) and the model seems to handle the UUIDs really well.

File: src/backend/chat/document_context_builder.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+
+    Rules:
+    - Reserve a ratio of max model context for full document inclusion.
+    - Keep all documents listed with an explicit accessibility status.
+    - Apply FIFO eviction on inlined documents when budget is exceeded.
+    """
+    if not text_attachments:
+        return ""
+
+    if not model_max_context:
```
Comment (providenz) [inline src/backend/chat/document_context_builder.py]: This path seems untested

File: src/backend/chat/document_context_builder.py [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+            "Model '%s' has no max_token_context; skipping full document inlining.",
+            model_hrid,
+        )
+        payload = _build_tool_only_payload(text_attachments)
+        return (
+            "List of documents attached to this conversation:\n"
+            f"{json.dumps(payload, ensure_ascii=False, indent=2)}"
+        )
+
+    if budget_ratio == 0:
```
Comment (providenz) [inline src/backend/chat/document_context_builder.py:256]: untested too

File: src/backend/chat/clients/pydantic_ai.py [hunk 16]
```diff
@@ -683,7 +708,38 @@ def _prepare_prompt(  # noqa: PLR0912  # pylint: disable=too-many-branches
...
+        return budget_ratio
+
+    async def _build_document_context_instruction(self) -> str:
+        budget_ratio = self._get_document_context_budget_ratio()
+        text_attachments = await sync_to_async(list)(
+            self.conversation.attachments.filter(content_type__startswith="text/").order_by(
+                "created_at", "id"
+            )
+        )
+        model_max_context = getattr(
```
Comment (providenz) [inline src/backend/chat/clients/pydantic_ai.py]: the getattr is unecessary, max_token_context is always here now (https://github.com/suitenumerique/conversations/pull/396/changes#diff-f62d84d03c0f3c349e96737dbcb62ac40ab7f307bf30070e1b4b31e8bd5e8ea9R177)

File: src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py [hunk 17]
```diff
@@ -172,13 +172,26 @@ def search(self, query: str, results_count: int = 4, **kwargs) -> RAGWebResults:
         Args:
             query (str): The search query.
             results_count (int): The number of results to return.
+            document_name (str, optional): If provided, only return search results from
+            the document whose title exactly matches this value.
```
Comment (providenz) [inline src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py]: nit: title/name

File: src/backend/chat/tests/clients/pydantic_ai/test_document_context_window.py [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+        "chat.clients.pydantic_ai.read_attachment_content",
+        fake_read_attachment_content,
+    )
+    monkeypatch.setattr(
+        "chat.clients.pydantic_ai.count_approx_tokens",
+        lambda _text: 1201,
+    )
+
+    instruction = async_to_sync(service._build_document_context_instruction)()  # pylint: disable=protected-access
+    assert '"title": "big"' in instruction
```
Comment (providenz) [inline src/backend/chat/tests/clients/pydantic_ai/test_document_context_window.py]: we can directly assert json shape

File: src/backend/chat/tests/clients/pydantic_ai/test_document_context_window.py [hunk 19]
```diff
@@ -0,0 +1,168 @@
+"""Tests for document context rolling window instructions."""
```
Comment (providenz) [inline src/backend/chat/tests/clients/pydantic_ai/test_document_context_window.py:1]: We could test with a more rela life approoach, without mocking and faking and use the rela tokenizer

File: src/backend/chat/tests/tools/test_document_search_rag.py [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+
+
+@pytest.mark.asyncio
+async def test_document_search_rag_forwards_document_id_to_backend(settings):
+    """The optional document_id argument must target and forward selected document."""
+    settings.RAG_DOCUMENT_SEARCH_BACKEND = (
+        "chat.agent_rag.document_rag_backends.albert_rag_backend.AlbertRagBackend"
+    )
+
+    mock_backend = mock.Mock()
```
Comment (providenz) [inline src/backend/chat/tests/tools/test_document_search_rag.py]: Instead of mocking and SimpleNamespace, we can test with real conv and attachments, and use respx to record albert  http calls

**Review by providenz — CHANGES_REQUESTED** [review]
Nice work, seems robust to me. 
- Some fixes needed in comments
- find backend should also be addressed
- as discussed, some question about tiktoken estimation accuracy with on openai model


## tests

- We need a few tests to ensure the model sees the inlinded content, not sure it is tested here.
- About the silent filter drop (see comment), it should also be tested 
- count_approx_tokens has no tests (no need to tets an axact match, but at least it provides a plausible estimation in different cases, non ascci, emojis, empty strings etc)

File: src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py (hunk 3 above)
Comment (camilleAND) [reply src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py]: This is to prevent case where the model used a document id as filter where it was in fact not needed. The Search will then retry but without any filter (like the legacy process). It can be useful to make sure some results are retrieved but can also output non-relevant chunks. We can choose to not do this and just tell the model the search was unsuccessful. The goal wad just to fallback to legacy way of doing rag search.

File: src/backend/chat/tools/document_summarize.py (hunk 4 above)
Comment (camilleAND) [reply src/backend/chat/tools/document_summarize.py]: Yes it was. Noted.

File: src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py (hunk 6 above)
Comment (camilleAND) [reply src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py]: Yes, but you're right, I'm changing this to not add metadata_filters at all when not needed for clarity.

File: src/backend/chat/document_context_builder.py (hunk 10 above)
Comment (camilleAND) [reply src/backend/chat/document_context_builder.py]: one attachment will just read info="first_uploaded_document" which is fine IMO

File: src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py [hunk 21]
```diff
@@ -216,23 +239,40 @@ async def asearch(self, query, results_count: int = 4, **kwargs) -> RAGWebResult
...
+            document_name (str, optional): If provided, only return search results from
+            the document whose name exactly matches this value.
             **kwargs: Additional arguments.
 
         Returns:
             RAGWebResults: The search results.
         """
         collection_ids = self.get_all_collection_ids()  # might raise RuntimeError
 
+        # Optional : document_names filter
```
Comment (maxenceh) [inline src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py]: we could create a `_build_quer` static method to avoid duplicating code

File: src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py [hunk 22]
```diff
@@ -172,28 +172,51 @@ def search(self, query: str, results_count: int = 4, **kwargs) -> RAGWebResults:
         Args:
             query (str): The search query.
             results_count (int): The number of results to return.
+            document_name (str, optional): If provided, only return search results from
```
Comment (maxenceh) [inline src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py]: why not adding `document_name` as an optional method param rather than getting it from kwargs?

File: src/backend/chat/clients/pydantic_ai.py [hunk 23]
```diff
@@ -687,7 +720,63 @@ def _prepare_prompt(  # noqa: PLR0912  # pylint: disable=too-many-branches
 
         return user_prompt[0], attachment_images, attachment_documents
 
-    def _setup_rag_tools(self) -> None:
+    def _get_document_context_budget_ratio(self) -> float:
```
Comment (maxenceh) [inline src/backend/chat/clients/pydantic_ai.py]: This could be a property in settings.py

File: src/backend/chat/clients/pydantic_ai.py [hunk 24]
```diff
@@ -687,7 +720,63 @@ def _prepare_prompt(  # noqa: PLR0912  # pylint: disable=too-many-branches
...
+        """Return a validated document-context ratio from settings."""
+        budget_ratio = settings.DOCUMENT_CONTEXT_BUDGET_RATIO
+        if budget_ratio < 0 or budget_ratio > 1:
+            logger.warning(
+                "Invalid DOCUMENT_CONTEXT_BUDGET_RATIO=%s, falling back to 0.5",
+                budget_ratio,
+            )
+            return 0.5
```
Comment (maxenceh) [inline src/backend/chat/clients/pydantic_ai.py]: There is a risk that the expected ratio is not applied without us knowing it.
Shouldn't we raise an error when ratio is <0 or >1?

File: src/backend/chat/tools/document_summarize.py [hunk 25]
```diff
@@ -79,19 +153,46 @@ async def document_summarize(  # pylint: disable=too-many-locals
...
 
         if not text_attachment:
             raise ModelCannotRetry(
                 "No text documents found in the conversation. "
                 "You must explain this to the user and ask them to provide documents."
             )
 
-        documents = [await read_document_content(doc) for doc in text_attachment]
+        if document_id is not None:
+            try:
```
Comment (maxenceh) [inline src/backend/chat/tools/document_summarize.py]: Same code as in document_search_rag.py, should we refactor with a function in utils.py?

Comment (maxenceh) [inline src/backend/chat/tools/document_summarize.py]: Same code as in document_search_rag.py, should we refactor with a function in utils.py?

File: src/backend/chat/clients/pydantic_ai.py [hunk 26]
```diff
@@ -166,6 +170,35 @@
 DOCUMENT_URL_PREFIX = "/media-key/"
 
 
+@functools.lru_cache(maxsize=1)
+def _get_token_encoding():
+    """Lazily load and cache the tiktoken encoding."""
+    return tiktoken.get_encoding("cl100k_base")
+
+
+def count_approx_tokens(text: str) -> int:
```
Comment (maxenceh) [inline src/backend/chat/clients/pydantic_ai.py]: why is this function defined here and not in `document_context_builder.p`?

File: src/backend/chat/document_context_builder.py [hunk 27]
```diff
@@ -0,0 +1,288 @@
...
+    docs: list[_DocumentEntry] = list(
+        await asyncio.gather(
+            *(
+                _load_document(index, attachment)
+                for index, attachment in enumerate(text_attachments, start=1)
+            )
+        )
+    )
+
+    inlined_total = _apply_fifo_inlining(docs, document_budget, conversation_id, logger)
```
Comment (maxenceh) [inline src/backend/chat/document_context_builder.py]: I find it a bit hard to understand that we modify `docs` here.
An option could be to have _apply_fifo_inlining returning an object:
```
@dataclasses.dataclass
class _DocumentEntryListing:
    """LLM-visible shape for the full attached-documents listing."""

    document_entries: list[_DocumentEntry]
    inlined_total: int
    inline_count: int
```

**Review by maxenceh — COMMENTED** [review]
Nice one! A few comment (open to discussion). 
I didn't review the tests yet.

File: src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py (hunk 22 above)
Comment (providenz) [reply src/backend/chat/agent_rag/document_rag_backends/albert_rag_backend.py]: signature comes from the abc, but yeah, cleaner imo

File: src/backend/chat/clients/pydantic_ai.py (hunk 26 above)
Comment (providenz) [reply src/backend/chat/clients/pydantic_ai.py]: is is used only here, but let's reduce pydantic line count :)

File: src/backend/chat/document_context_builder.py (hunk 27 above)
Comment (providenz) [reply src/backend/chat/document_context_builder.py]: it was extracted from the main function to make sonarkube happy, the edit in place is a bit smelly

File: src/backend/chat/tools/document_summarize.py (hunk 25 above)
Comment (providenz) [reply src/backend/chat/tools/document_summarize.py]: ok

File: src/backend/chat/clients/pydantic_ai.py [hunk 28]
```diff
@@ -729,7 +732,49 @@ def _prepare_prompt(  # noqa: PLR0912  # pylint: disable=too-many-branches
...
-    def _setup_rag_tools(self) -> None:
+    async def _build_document_context_instruction(self) -> str:
+        budget_ratio = settings.DOCUMENT_CONTEXT_BUDGET_RATIO
+        security_buffer_tokens = settings.DOCUMENT_CONTEXT_SECURITY_BUFFER_TOKENS
+        text_attachments = await sync_to_async(list)(
+            self.conversation.attachments.filter(
+                content_type__startswith=TEXT_MIME_PREFIX
+            ).order_by("created_at", "id")
+        )
+        model_max_context = self.conversation_agent.configuration.max_token_context
```
Comment (maxenceh) [inline src/backend/chat/clients/pydantic_ai.py]: This is called `model_max_context` here and `max_token_context` in config.
Might worth using the same name

File: src/backend/chat/document_context_builder.py [hunk 29]
```diff
@@ -0,0 +1,329 @@
...
+
+
+async def build_document_context_instruction(  # noqa: PLR0913 # pylint: disable=too-many-arguments,too-many-locals
+    *,
+    conversation_id: str,
+    text_attachments: Sequence[models.ChatConversationAttachment],
+    model_hrid: str,
+    model_max_context: int | None,
+    budget_ratio: float,
+    security_buffer_tokens: int,
```
[the remark this answers is not in the document]
Comment (providenz) [reply src/backend/chat/document_context_builder.py:220]: enforced at the settings level

**Review by providenz — APPROVED** [review]
